Add nursery fixture

I'm currently seeing a big limitation to trio async fixtures: there is no nursery available when they are run.

Considering a usecase when you want to test the connection to a server, you have to:
1) start the server
2) connect to the server
3) send a request to the server and check the result
4) close the server

Without nursery, there is no way to achieve 1) from a fixture, so this means the test would look like this:

```python
async def test_server_connection():

    async def actual_test(listerner):
        connection = await connect_to_server(listerner)  # step 2)
        ret = await connection.ping()  # step 3)
        assert ret == 'ok'
        nursery.cancel_scope.cancel()  # step 4)

    async with trio.open_nursery() as nursery:
        listerners = nursery.start(trio.serve_tcp, my_server_handle_client, 0)  # step 1)
        nursery.start_soon(actual_test, listerners[0])
```

This is pretty cumbersome (especially with multiple tests, each one needing to start the server).

A solution to this trouble could be to write a decorator for the test:

```python
async def with_server(fn):

    @wraps(fn)
    async def wrapper(*args, **kwargs):

        async def runtest_then_cancel_scope(listerner):
            await fn(listerner, *args, **kwargs)
            nursery.cancel_scope.cancel()  # step 4)

        async with trio.open_nursery() as nursery:
            listeners = nursery.start(trio.serve_tcp, my_server_handle_client, 0)  # step 1)
            nursery.start_soon(runtest_then_cancel_scope, listerners[0])


@with_server
async def test_server_connection(server):
    connection = await connect_to_server(server)  # step 2)
    ret = await connection.ping()  # step 3)
    assert ret == 'ok'
```

While the test looks much better, boilerplates are still pretty cumbersome. On the top of that this code is wrong because `@wraps` will pass all the functions parameters to the parent caller which in turn will consider them as fixtures to find. However the first parameter `server` is in fact injected by the decorator itself ! The work around is to create a custom `wraps` function that discard the first argument before doing the decoration... This is getting out of hand pretty quickly !

To solve all this, I'm thinking the user test function could be run into a nursery, this way we could provide a `nursery` fixture to access it, allowing to easily create fixture responsible to start a server.

```python
@pytest.fixture
async def server(nursery):
    listeners = await nursery.start(trio.serve_tcp, my_server_handle_client, 0)  # step 1)
    return listeners[0]


async def test_server_connection(server):
    connection = await connect_to_server(server) # step 2)
    ret = await connection.ping()  # step 3)
    assert ret == 'ok'
```

Finally we would call `nursery.cancel_scope.cancel()` once the teardown of the fixture have been done (which would achieve step 4 automatically).

This seems much more elegant and straightforward, however I'm still unsure if canceling this scope automatically is such a good idea (this could lead to normally infinite loop that would end up hidden).
Another solution would be to take advantage of the async yield fixture to do stop the server during teardown of the fixture:

```python
@pytest.fixture
async def server(nursery):
    with trio.open_cancel_scope():
        listeners = await nursery.start(trio.serve_tcp, handle_client, 0)
        yield listeners[0]
```

However this doesn't seems to work:
```
../../.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py:1502: in run_impl
    msg = task.coro.send(next_send)
pytest_trio/plugin.py:96: in teardown
    await self._teardown()
pytest_trio/plugin.py:119: in _teardown
    await self.agen.asend(None)
pytest_trio/_tests/test_try.py:14: in server
    yield listeners[0]
/usr/local/lib/python3.6/contextlib.py:89: in __exit__
    next(self.gen)
../../.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py:206: in open_cancel_scope
    scope._remove_task(task)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <cancel scope object at 0x7f1ee62c41d0>, task = <Task 'pytest_trio.plugin._setup_async_fixtures_in.<locals>._resolve_and_update_deps' at 0x7f1ee62c4208>

    def _remove_task(self, task):
        with self._might_change_effective_deadline():
>           self._tasks.remove(task)
E           KeyError: <Task 'pytest_trio.plugin._setup_async_fixtures_in.<locals>._resolve_and_update_deps' at 0x7f1ee62c4208>

../../.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py:165: KeyError
```

I'm not sure if this bug is due to a conflict between `open_nursery` and the `open_cancel_scope`  over the `server_tcp` coroutine or a more generic bug caused by how an async yield fixture is created inside the `setup` coroutine but finished consumed into a `teardown` one...


@njsmith I would be really interested to have your point of view on this
I've [created a branch](https://github.com/touilleMan/pytest-trio/tree/nursery-fixture) implementing this feature, you can have a look at the [test crashing](https://github.com/touilleMan/pytest-trio/blob/nursery-fixture/pytest_trio/_tests/test_nursery_fixture.py)

> This has the problem that it violates the "parenting is a full time job" rule

Yep, that was the reason why I didn't want to do it this way...

> I'm currently working on fixing Trio to get rid of the parenting-is-a-full-time-job rule.

Cool ! do you have a link with more info about this ?

> In your code, I think you meant to write something like this?

Yeah I wasn't sure the `cancel_scope.cancel()` was mandatory when using °open_cancel_scope()` (I though the cancel was automatic once reached the `__exit__`)

> I feel sad about losing the cool recursive/parallel fixture setup, but maybe we need to switch to something simpler, where we instantiate the fixtures serially in the main task, and then clean them up in serial in the opposite order.

I've created PR #26 that remove the parallel fixture setup/teardown. This way the nursery creation from within the fixture works well. However this is not perfect for error handling coming from the coroutine started within the nursery:

Considering:
```python
async def handle_client(stream):
    while True:
        raise RuntimeError('fooo')
        buff = await stream.receive_some(4)
        await stream.send_all(buff)
```

We end up with:
```
E                   trio.TrioInternalError: internal error in trio - please file a bug!

../../.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py:1394: TrioInternalError
============================================================================================= 1 failed in 0.11 seconds ==============================================================================================
Exception ignored in: <bound method Nursery.__del__ of <trio._core._run.Nursery object at 0x7fa2091536a0>>
Traceback (most recent call last):
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py", line 522, in __del__
AssertionError: 
Exception ignored in: <bound method AsyncGenerator.__del__ of <async_generator.impl.AsyncGenerator object at 0x7fa209153080>>
Traceback (most recent call last):
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/async_generator/impl.py", line 310, in __del__
RuntimeError: partially-exhausted async_generator garbage collected
Exception ignored in: <bound method Nursery.__del__ of <trio._core._run.Nursery object at 0x7fa209133cf8>>
Traceback (most recent call last):
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py", line 522, in __del__
AssertionError: 
Exception ignored in: <async_generator object server at 0x7fa2095edc78>
Traceback (most recent call last):
  File "/home/emmanuel/projects/pytest-trio/pytest_trio/_tests/test_nursery_fixture.py", line 17, in server
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_util.py", line 120, in __aexit__
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/async_generator/impl.py", line 264, in athrow
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/async_generator/impl.py", line 277, in _do_it
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/async_generator/impl.py", line 187, in __next__
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/async_generator/impl.py", line 199, in _invoke
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py", line 320, in open_nursery
  File "/usr/local/lib/python3.6/contextlib.py", line 100, in __exit__
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py", line 206, in open_cancel_scope
  File "/home/emmanuel/.local/share/virtualenvs/pytest-trio-fHSyI99q/lib/python3.6/site-packages/trio/_core/_run.py", line 165, in _remove_task
KeyError: <Task 'pytest_trio.plugin._trio_test_runner_factory.<locals>._bootstrap_fixture_and_run_test' at 0x7fa209133a20>
```

So I'm still hoping for a better way to do that...

--- a/pytest_trio/_tests/test_async_yield_fixture.py
+++ b/pytest_trio/_tests/test_async_yield_fixture.py
@@ -224,3 +224,40 @@ def test_async_yield_fixture_with_multiple_yields(testdir):
     # TODO: should trigger error instead of failure
     # result.assert_outcomes(error=1)
     result.assert_outcomes(failed=1)
+
+
+@pytest.mark.skipif(sys.version_info < (3, 6), reason="requires python3.6")
+def test_async_yield_fixture_with_nursery(testdir):
+
+    testdir.makepyfile(
+        """
+        import pytest
+        import trio
+
+
+        async def handle_client(stream):
+            while True:
+                buff = await stream.receive_some(4)
+                await stream.send_all(buff)
+
+
+        @pytest.fixture
+        async def server():
+            async with trio.open_nursery() as nursery:
+                listeners = await nursery.start(trio.serve_tcp, handle_client, 0)
+                yield listeners[0]
+                nursery.cancel_scope.cancel()
+
+
+        @pytest.mark.trio
+        async def test_actual_test(server):
+            stream = await trio.testing.open_stream_to_socket_listener(server)
+            await stream.send_all(b'ping')
+            rep = await stream.receive_some(4)
+            assert rep == b'ping'
+    """
+    )
+
+    result = testdir.runpytest()
+
+    result.assert_outcomes(passed=1)

--- a/pytest_trio/plugin.py
+++ b/pytest_trio/plugin.py
@@ -39,24 +39,17 @@ def _trio_test_runner_factory(item):
 async def _setup_async_fixtures_in(deps):
     resolved_deps = {**deps}
 
-    async def _resolve_and_update_deps(afunc, deps, entry):
-        deps[entry] = await afunc()
+    for depname, depval in resolved_deps.items():
+        if isinstance(depval, BaseAsyncFixture):
+            resolved_deps[depname] = await depval.setup()
 
-    async with trio.open_nursery() as nursery:
-        for depname, depval in resolved_deps.items():
-            if isinstance(depval, BaseAsyncFixture):
-                nursery.start_soon(
-                    _resolve_and_update_deps, depval.setup, resolved_deps,
-                    depname
-                )
     return resolved_deps
 
 
 async def _teardown_async_fixtures_in(deps):
-    async with trio.open_nursery() as nursery:
-        for depval in deps.values():
-            if isinstance(depval, BaseAsyncFixture):
-                nursery.start_soon(depval.teardown)
+    for depval in deps.values():
+        if isinstance(depval, BaseAsyncFixture):
+            await depval.teardown()
 
 
 class BaseAsyncFixture:

